src/repositories/crates/api/v1/mod.rs

```diff
@@ -20,25 +23,44 @@ async fn download(
         if let Some(crate_version) = info.versions.iter().find(|e| e.num == version) {
             let link = &crate_version.dl_path;
 
-            if let Ok(mut upstream) = policy
-                .client
-                .get(format!("https://crates.io/{link}"))
-                .send()
-                .await
-            {
+            let url = format!("https://crates.io/{link}");
+            if let Ok(mut upstream) = policy.client.get(url.clone()).send().await {
                 if let Ok(payload) = upstream.body().limit(20_000_000).await {
                     let digest = sha256::digest(payload.as_ref());
                     let uuids = search(digest.clone()).await;
                     println!("{crate_name} {version} = {digest} {uuids:?}");
 
-                    let mut response = HttpResponse::Ok();
-                    if let Some(v) = upstream.headers().get(header::CONTENT_TYPE) {
-                        response.append_header((header::CONTENT_TYPE, v));
+                    let id = ArtifactIdentifier::Crate {
+                        name: crate_name.clone(),
+                    };
+                    let context = Context::new(
+                        url,
+                        sha256::digest(payload.as_ref()), // todo: double check this
+                        id,
+                        crates.scope.to_owned(),
+                    );
+
+                    match policy.evaluate(&context).await {
+                        Ok(None) => {
+                            log::info!("Policy evaluation success");
+                            let mut response = HttpResponse::Ok();
+                            if let Some(v) = upstream.headers().get(header::CONTENT_TYPE) {
+                                response.append_header((header::CONTENT_TYPE, v));
+                            }
+                            let disposition =
+                                format!("attachment; filename=\"{crate_name}-{version}.crate\"");
+                            response.append_header((header::CONTENT_DISPOSITION, disposition));
+                            return response.body(payload);
```

Comment:
Personally, I prefer omitting `return` unless the code is clearer using an "early exit".

```suggestion
                            response.body(payload)
```

Reply:
After looking at this for a bit, it is an "early exit" with the way the code is structured currently. I've been reworking it to look more like the maven `proxy()` function, but in this case I think it may be making the logic harder to follow as we're returning many types of errors and have so many nested matches.

Reply:
All this to say, still working on it, as I'd like the two to look more similar rather than less.